src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala

```diff
@@ -37,6 +40,31 @@ class CampaignResource(extractAuth: Directive1[AuthedNamespaceScope],
     } yield campaignId
   }
 
+  /**
+    * Create and immediately launch a retry-campaign for all the devices that were processed by `mainCampaign` and
+    * failed with a code `failureCode`. We assume there is at least one failed device. Hence, if there were not any
+    * such devices in `mainCampaign`, throw an error.
+    */
+  def retryFailedDevices(ns: Namespace, mainCampaign: Campaign, request: RetryFailedDevices): Future[Unit] =
+    campaigns.fetchFailedDevices(mainCampaign.id, request.failureCode).map(_.toSeq).flatMap {
+      case Nil =>
+        Future.failed(MissingFailedDevices(request.failureCode))
+
+      case deviceIds =>
+        // FIXME we just need to create this dummy group until we get totally rid of the groups.
+        val retryGroup = GroupId(UUID.fromString("0000-00-00-00-000000"))
```

Comment:
The latest commits from #106 are loosen the groups type from `NonEmptyList` to just `Set`, so you won't need a fake group here and in `CreateRetryCampaign` class when you rebase on top of it.

Reply:
Nevermind, just saw that you already did it